derive/src/lib.rs

```diff
@@ -76,7 +76,29 @@ fn generate(input: TokenStream2) -> Result<TokenStream2> {
 fn generate_type(input: TokenStream2) -> Result<TokenStream2> {
     let mut ast: DeriveInput = syn::parse2(input.clone())?;
 
-    let scale_info = crate_name_ident("scale-info")?;
+    // Find the name given to the `scale-info` crate in the context we run in.
+    // If scale-info is not among the dependencies then we must be deriving
+    // types for the scale-info crate itself, in which case we need to rename
+    // "self" to something, so the object paths keep working.
+    let (scale_info, import_self_as_scale_info) = {
+        let actual_crate_name = proc_macro_crate::crate_name("scale-info");
+        if let Err(e) = actual_crate_name {
+            if e.starts_with("Could not find `scale-info`") {
+                (Ident::new("_scale_info", Span::call_site()), true)
+            } else {
+                return Err(syn::Error::new(Span::call_site(), e))
+            }
+        } else {
+            (
+                Ident::new(
+                    &actual_crate_name.expect("Checked Err above; qed"),
```

Comment:
Could use better error message. I find `unwrap_or_else(|p| panic!(..))` to deliver far better error messages than `expect`.

Reply:
Don't you think that's overkill? This could be `unwrap()` – the error case is checked a few lines above so the comment is for the code reader trained to balk at `unwrap()`s.